src/ResultSet.php

```diff
@@ -3,45 +3,107 @@
 namespace ipl\Orm;
 
 use ArrayIterator;
+use BadMethodCallException;
+use Generator;
 use Iterator;
 use Traversable;
 
+/**
+ * Dataset containing database rows
+ */
 class ResultSet implements Iterator
 {
+    /** @var ArrayIterator */
     protected $cache;
 
     /** @var bool Whether cache is disabled */
     protected $isCacheDisabled = false;
 
+    /** @var Generator */
     protected $generator;
 
+    /** @var ?int */
     protected $limit;
 
+    /** @var ?int */
     protected $position;
 
-    public function __construct(Traversable $traversable, $limit = null)
+    /** @var ?int */
+    protected $offset;
+
+    /** @var ?int */
+    protected $pageSize;
+
+    /**
+     * Construct the ResultSet object
+     *
+     * @param Traversable $traversable
+     * @param ?int $limit
+     * @param ?int $offset
+     */
+    public function __construct(Traversable $traversable, $limit = null, $offset = null)
     {
         $this->cache = new ArrayIterator();
         $this->generator = $this->yieldTraversable($traversable);
         $this->limit = $limit;
+        $this->offset = $offset;
+    }
+
+    /**
+     * Get the current page number
+     *
+     * Returns the current page, calculated by the {@see self::$position position}, {@see self::$offset offset}
+     * and the {@see self::$pageSize page size}
+     *
+     * @return int page
+     * @throws BadMethodCallException if no {@see self::$pageSize page size} has been provided
+     */
+    public function getCurrentPage(): int
+    {
+        if ($this->pageSize) {
+            $offset = $this->offset ?: 0;
+            $position = $this->position + 1;
+            if ($position && ($position + $offset) > $this->pageSize) {
```

Comment:
If you had to check if `$permission` is non-zero/non-null, why is the addition above now safe and this check still there?

Reply:
That was indeed an error that I did not catch when adding the inline position variable